pkg/kubelet/images/puller.go

```diff
@@ -34,42 +37,102 @@ type pullResult struct {
 }
 
 type imagePuller interface {
-	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan<- pullResult, *runtimeapi.PodSandboxConfig)
+	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan pullResult, *runtimeapi.PodSandboxConfig, v1.PullPolicy)
 }
 
 var _, _ imagePuller = &parallelImagePuller{}, &serialImagePuller{}
 
 type parallelImagePuller struct {
 	imageService kubecontainer.ImageService
 	tokens       chan struct{}
+	pulls        sync.Map
+}
+
+type pullBroker struct {
+	subscribers sync.Map
+}
+
+func (p *pullBroker) subscribe(ch chan pullResult) {
+	p.subscribers.Store(ch, struct{}{})
+}
+
+func (p *pullBroker) publish(res pullResult) {
+	p.subscribers.Range(func(k, v any) bool {
+		ch, ok := k.(chan pullResult)
+		if !ok {
+			return false
+		}
+
+		select {
+		case ch <- res:
+		default:
+		}
+
+		return true
+	})
 }
 
 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
 	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-		return &parallelImagePuller{imageService, nil}
+		return &parallelImagePuller{imageService, nil, sync.Map{}}
 	}
-	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls)}
+	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
 }
 
-func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan<- pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
+func pullKey(spec kubecontainer.ImageSpec) string {
+	res := []string{}
+	for _, annotation := range spec.Annotations {
+		res = append(res, annotation.Name+"-"+annotation.Value)
+	}
+	sort.Strings(res)
+	if spec.RuntimeHandler != "" {
+		res = append([]string{spec.RuntimeHandler}, res...)
+	}
+	if spec.Image != "" {
+		res = append([]string{spec.Image}, res...)
+	}
+	return strings.Join(res, "-")
+}
+
+func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig, pullPolicy v1.PullPolicy) {
 	go func() {
+		key := pullKey(spec)
+
+		// A pull policy of 'Always' will result in calling CRI's PullImage
+		// every time.
+		if pullPolicy != v1.PullAlways {
+			// Reuse existing pull broker if available
+			if v, ok := pip.pulls.Load(key); ok {
```

Comment:
my understanding of this code that there is no synchronization between Store and Load. So if two job pods scheduled at the same time, we still may end up with both images taking on of the tokens. Are we assuming that the race like this will be rare?

Reply:
This point seems to be related to the unit test (`TestPullImageParallel`) being flaky. I'm not sure if using a mutex for synchronization between Store and Load is the correct approach, but the following changes made the test pass.

```console
❯ stress ./images.test -test.run TestPullImageParallel
(...)
7m55s: 2256 runs so far, 0 failures
```

<details>
<summary>diff</summary>


```diff
diff --git a/pkg/kubelet/images/puller.go b/pkg/kubelet/images/puller.go
index 30d64336e85..d0b1b686d49 100644
--- a/pkg/kubelet/images/puller.go
+++ b/pkg/kubelet/images/puller.go
@@ -46,6 +46,7 @@ type parallelImagePuller struct {
        imageService kubecontainer.ImageService
        tokens       chan struct{}
        pulls        sync.Map
+       inflightPull sync.Mutex
 }

 type pullBroker struct {
@@ -74,9 +75,9 @@ func (p *pullBroker) publish(res pullResult) {

 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
        if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-               return &parallelImagePuller{imageService, nil, sync.Map{}}
+               return &parallelImagePuller{imageService, nil, sync.Map{}, sync.Mutex{}}
        }
-       return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
+       return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}, sync.Mutex{}}
 }

 func pullKey(spec kubecontainer.ImageSpec) string {
@@ -97,7 +98,7 @@ func pullKey(spec kubecontainer.ImageSpec) string {
 func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig, pullPolicy v1.PullPolicy) {
        go func() {
                key := pullKey(spec)
-
+               pip.inflightPull.Lock()
                // A pull policy of 'Always' will result in calling CRI's PullImage
                // every time.
                if pullPolicy != v1.PullAlways {
@@ -105,6 +106,7 @@ func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontaine
                        if v, ok := pip.pulls.Load(key); ok {
                                existingPullBroker, _ := v.(*pullBroker)
                                existingPullBroker.subscribe(pullChan)
+                               pip.inflightPull.Unlock()
                                return
                        }
                }
@@ -118,6 +120,7 @@ func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontaine
                br := &pullBroker{}
                br.subscribe(pullChan)
                pip.pulls.Store(key, br)
+               pip.inflightPull.Unlock()

                startTime := time.Now()
                imageRef, err := pip.imageService.PullImage(ctx, spec, pullSecrets, podSandboxConfig)
diff --git a/pkg/kubelet/images/puller_test.go b/pkg/kubelet/images/puller_test.go
index 3af9f6ac138..d7e26910557 100644
--- a/pkg/kubelet/images/puller_test.go
+++ b/pkg/kubelet/images/puller_test.go
@@ -18,6 +18,7 @@ package images

 import (
        "context"
+       "sync"
        "sync/atomic"
        "testing"
        "time"
@@ -26,6 +27,7 @@ import (
        v1 "k8s.io/api/core/v1"
        kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
        ctest "k8s.io/kubernetes/pkg/kubelet/container/testing"
+       "k8s.io/utils/ptr"
 )

 func TestPullKey(t *testing.T) {
@@ -76,29 +78,103 @@ func TestPullKey(t *testing.T) {
 }

 func TestPullImageParallel(t *testing.T) {
+       tests := []struct {
+               name                  string
+               images                []kubecontainer.ImageSpec
+               imagePullPolicy       v1.PullPolicy
+               expectedImagePullCall uint32
+       }{
+               {
+                       name: "single image",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                       },
+                       imagePullPolicy:       v1.PullIfNotPresent,
+                       expectedImagePullCall: 1,
+               },
+               {
+                       name: "multiple same images",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                               {Image: "image1"},
+                               {Image: "image1"},
+                       },
+                       imagePullPolicy:       v1.PullIfNotPresent,
+                       expectedImagePullCall: 1,
+               },
+               {
+                       name: "multiple different images",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                               {Image: "image2"},
+                               {Image: "image1"},
+                               {Image: "image2"},
+                       },
+                       imagePullPolicy:       v1.PullIfNotPresent,
+                       expectedImagePullCall: 2,
+               },
+               {
+                       name: "all different images",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                               {Image: "image2"},
+                               {Image: "image3"},
+                               {Image: "image4"},
+                               {Image: "image5"},
+                       },
+                       imagePullPolicy:       v1.PullIfNotPresent,
+                       expectedImagePullCall: 5,
+               },
+               {
+                       name: "multiple same images with Always pull policy",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                               {Image: "image1"},
+                               {Image: "image1"},
+                       },
+                       imagePullPolicy:       v1.PullAlways,
+                       expectedImagePullCall: 3,
+               },
+               {
+                       name: "multiple different images with Always pull policy",
+                       images: []kubecontainer.ImageSpec{
+                               {Image: "image1"},
+                               {Image: "image2"},
+                               {Image: "image3"},
+                               {Image: "image2"},
+                               {Image: "image1"},
+                       },
+                       imagePullPolicy:       v1.PullAlways,
+                       expectedImagePullCall: 5,
+               },
+       }
+
        const (
-               parallelPulls = 5
-               delay         = 10 * time.Millisecond
-       )
-       var (
-               maxParallelPulls int32 = parallelPulls
-               results          atomic.Uint32
+               delay             = 10 * time.Millisecond
+               parallelPulls int = 3
        )
+       maxParallelPulls := ptr.To(int32(parallelPulls))
+       var wg sync.WaitGroup
+       for _, test := range tests {
+               var results atomic.Uint32
+               fake := &ctest.FakeRuntime{T: t, DelayImagePulls: delay * time.Duration(parallelPulls)}
+               p := newParallelImagePuller(fake, maxParallelPulls)

-       fake := &ctest.FakeRuntime{T: t, DelayImagePulls: parallelPulls * delay}
-       p := newParallelImagePuller(fake, &maxParallelPulls)
+               t.Run(test.name, func(t *testing.T) {
+                       for _, image := range test.images {
+                               wg.Add(1)
+                               resCh := make(chan pullResult)
+                               p.pullImage(context.TODO(), image, nil, resCh, nil, test.imagePullPolicy)
+                               go func() {
+                                       <-resCh
+                                       results.Add(1)
+                                       wg.Done()
+                               }()
+                       }

-       for range parallelPulls {
-               resCh := make(chan pullResult)
-               p.pullImage(context.TODO(), kubecontainer.ImageSpec{Image: "image"}, nil, resCh, nil, v1.PullIfNotPresent)
-               go func() {
-                       <-resCh
-                       results.Add(1)
-               }()
-               time.Sleep(delay)
+                       wg.Wait()
+                       fake.AssertCallCounts("PullImage", int(test.expectedImagePullCall))
+                       require.EqualValues(t, uint32(len(test.images)), results.Load())
+               })
        }
-       time.Sleep(delay)
-
-       fake.AssertCallCounts("PullImage", 1)
-       require.EqualValues(t, results.Load(), parallelPulls)
 }
```

</details>

Reply:
I fixed the race between load and store by using the `LoadOrStore` API. The test case is still tricky from a timing perspective, we have to ensure that `PullImage` takes some time (like in the real world scenario) so that we can actually subscribe to an existing pull.